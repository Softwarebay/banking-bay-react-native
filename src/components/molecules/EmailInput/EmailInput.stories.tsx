import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';

import { ViewStory } from 'utils/storybook/ViewStory';

import { EmailInput } from './EmailInput';

storiesOf('Atoms/EmailInput', module)
	.addDecorator((getStory) => <ViewStory>{getStory()}</ViewStory>)
	.add('Valid', () => (
		<EmailInput
			isValid={boolean('isValid', true)}
			placeholder={text('placeholder', 'Enter your email')}
			testID={text('testID', 'EmailInput-test-id')}
		/>
	))
	.add('Not valid', () => (
		<EmailInput
			isValid={boolean('isValid', false)}
			placeholder={text('placeholder', 'Enter your email')}
			testID={text('testID', 'EmailInput-test-id')}
		/>
	));
