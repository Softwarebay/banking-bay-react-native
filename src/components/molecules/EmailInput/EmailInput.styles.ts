import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
	container: {
		height: 50,
		paddingHorizontal: 20,
		paddingVertical: 14,
	},
	icon: {
		right: 0,
	},
	inputField: {
		flex: 1,
		marginRight: 20,
	},
});
