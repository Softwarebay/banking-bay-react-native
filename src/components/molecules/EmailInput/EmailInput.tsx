import { Dispatch, forwardRef, SetStateAction } from 'react';
import { StyleProp, TextStyle } from 'react-native';

import { Card, Icon, InputField, InputFieldHandle } from 'components/atoms';

import { colors } from 'theme';

import { styles } from './EmailInput.styles';

interface Props {
	placeholder: string;
	testID: string;
	style?: StyleProp<TextStyle>;
	isValid: boolean;
	validate?: (text: string) => boolean;
	setIsValid?: Dispatch<SetStateAction<boolean>>;
}

export const EmailInput = forwardRef<InputFieldHandle, Props>(
	({ placeholder, style, validate, isValid, setIsValid, testID }, ref) => {
		return (
			<Card style={[styles.container, style]}>
				<InputField
					ref={ref}
					keyboardType="email-address"
					placeholder={placeholder}
					setIsValid={setIsValid}
					style={styles.inputField}
					testID={testID}
					validate={validate}
				/>
				{isValid && (
					<Icon
						color={colors.mainDark}
						name="check"
						testID={`icon-${testID}`}
					/>
				)}
			</Card>
		);
	}
);
