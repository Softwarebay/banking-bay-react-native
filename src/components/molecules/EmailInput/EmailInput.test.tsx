import {
	fireEvent,
	render,
	screen,
	within,
} from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { EmailInput } from './EmailInput';

import type { Props } from './EmailInput';

describe('EmailInput', () => {
	const props: Omit<Props, 'isValid'> = {
		testID: 'email-input-test-id',
		placeholder: 'Enter email',
		style: {
			color: 'red',
		},
	};

	it('should match snapshot - input is valid', () => {
		const tree = testRenderer
			.create(<EmailInput {...props} isValid />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should match snapshot - input is not valid', () => {
		const tree = testRenderer
			.create(<EmailInput {...props} isValid={false} />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should show typed in text', () => {
		render(<EmailInput {...props} isValid={false} />);

		const emailInputField = screen.getByTestId(
			'input-field-email-input-test-id'
		);

		expect(emailInputField.props).toHaveProperty('secureTextEntry', false);
	});

	it('should display placeholder', () => {
		render(<EmailInput {...props} isValid={false} />);

		const emailInputField = screen.getByTestId(
			'input-field-email-input-test-id'
		);

		expect(emailInputField.props).toHaveProperty('placeholder', 'Enter email');
	});

	it('should render check icon if input is valid', () => {
		render(<EmailInput {...props} isValid />);

		const emailInput = screen.getByTestId('email-input-test-id');
		const checkIcon = within(emailInput).getByTestId('check-icon-test-id');

		expect(emailInput.children).toHaveLength(2);
		expect(checkIcon).not.toBeNull();
	});

	it("shouldn't render check icon if input is not valid", () => {
		render(<EmailInput {...props} isValid={false} />);

		const emailInput = screen.getByTestId('email-input-test-id');
		const checkIcon = within(emailInput).queryByTestId('check-icon-test-id');

		expect(emailInput.children).toHaveLength(1);
		expect(checkIcon).toBeNull();
	});

	it('should handle type events', () => {
		render(<EmailInput {...props} isValid={false} />);

		const emailInputField = screen.getByTestId(
			'input-field-email-input-test-id'
		);

		fireEvent.changeText(emailInputField, 'test@test.com');

		expect(emailInputField.props).toHaveProperty('value', 'test@test.com');
	});
});
