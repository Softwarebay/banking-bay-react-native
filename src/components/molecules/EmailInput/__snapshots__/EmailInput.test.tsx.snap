// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailInput should match snapshot - input is not valid 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
      },
      Object {
        "flexDirection": "row",
      },
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 10,
          "shadowColor": "#161966",
          "shadowOffset": Object {
            "height": 6,
            "width": 0,
          },
          "shadowOpacity": 0.08,
          "shadowRadius": 16,
        },
        Array [
          Object {
            "height": 55,
            "paddingHorizontal": 20,
            "paddingVertical": 14,
          },
          Object {
            "color": "red",
          },
        ],
      ],
    ]
  }
  testID="email-input-test-id"
>
  <TextInput
    autoCorrect={false}
    keyboardType="email-address"
    onChangeText={[Function]}
    placeholder="Enter email"
    secureTextEntry={false}
    style={
      Array [
        Object {
          "color": "#1B1D4D",
          "fontFamily": "MulishRegular",
          "fontSize": 15,
        },
        Object {
          "flex": 1,
          "marginRight": 20,
        },
      ]
    }
    testID="input-field-email-input-test-id"
    value=""
  />
</View>
`;

exports[`EmailInput should match snapshot - input is valid 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
      },
      Object {
        "flexDirection": "row",
      },
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 10,
          "shadowColor": "#161966",
          "shadowOffset": Object {
            "height": 6,
            "width": 0,
          },
          "shadowOpacity": 0.08,
          "shadowRadius": 16,
        },
        Array [
          Object {
            "height": 55,
            "paddingHorizontal": 20,
            "paddingVertical": 14,
          },
          Object {
            "color": "red",
          },
        ],
      ],
    ]
  }
  testID="email-input-test-id"
>
  <TextInput
    autoCorrect={false}
    keyboardType="email-address"
    onChangeText={[Function]}
    placeholder="Enter email"
    secureTextEntry={false}
    style={
      Array [
        Object {
          "color": "#1B1D4D",
          "fontFamily": "MulishRegular",
          "fontSize": 15,
        },
        Object {
          "flex": 1,
          "marginRight": 20,
        },
      ]
    }
    testID="input-field-email-input-test-id"
    value=""
  />
  <SvgMock
    color="#1B1D4D"
    height={16}
    testID="check-icon-test-id"
    width={16}
  />
</View>
`;
